itest/lnd_sendonion_test.go

```diff
@@ -0,0 +1,143 @@
+	defer ht.CloseChannel(bob, chanPointBob)
+
+	// Next, we'll create Carol and establish a channel to from her to
+	// Dave.
+	ht.FundCoins(btcutil.SatoshiPerBitcoin, carol)
+	chanPointCarol := ht.OpenChannel(
+		carol, dave, lntest.OpenChannelParams{Amt: chanAmt},
+	)
+	defer ht.CloseChannel(carol, chanPointCarol)
+
+	// Make sure Alice knows the channel between Bob and Carol.
+	ht.AssertTopologyChannelOpen(alice, chanPointBob)
+	ht.AssertTopologyChannelOpen(alice, chanPointCarol)
+
+	const (
+		numPayments = 1
+		paymentAmt  = 10000
+	)
+
+	// Request an invoice from Dave so he is expecting payment.
+	_, rHashes, invoices := ht.CreatePayReqs(dave, paymentAmt, numPayments)
+	var preimage lntypes.Preimage
+	copy(preimage[:], invoices[0].RPreimage)
+
+	// Query for routes to pay from Alice to Dave.
+	routesReq := &lnrpc.QueryRoutesRequest{
+		PubKey: dave.PubKeyStr,
+		Amt:    paymentAmt,
+		// AmtMsat: paymentAmt,
+	}
+	routes := alice.RPC.QueryRoutes(routesReq)
+	route := routes.Routes[0]
+	finalHop := route.Hops[len(route.Hops)-1]
+	finalHop.MppRecord = &lnrpc.MPPRecord{
+		PaymentAddr:  invoices[0].PaymentAddr,
+		TotalAmtMsat: int64(lnwire.NewMSatFromSatoshis(paymentAmt)),
+	}
+
+	// Construct an onion for the route from Alice to Dave.
+	paymentHash := rHashes[0]
+	onionReq := &routerrpc.BuildOnionRequest{
+		Route:       route,
+		PaymentHash: paymentHash,
+	}
+	onionResp := alice.RPC.BuildOnion(onionReq)
+
+	// Dispatch a payment via the SendOnion RPC.
+	firstHop := bob.PubKey
+	sendReq := &routerrpc.SendOnionRequest{
+		FirstHopPubkey: firstHop[:],
+		Amount:         route.TotalAmtMsat,
+		Timelock:       route.TotalTimeLock,
+		PaymentHash:    paymentHash,
+		OnionBlob:      onionResp.OnionBlob,
+		AttemptId:      0,
+	}
+
+	// NOTE(calvin): We may want our wrapper RPC client to allow errors
+	// through so that we can make some assertions about them in various
+	// scenarios.
+	// resp, err := alice.RPC.SendOnion(onionReq)
+	// require.NoError(ht, err, "unable to send payment via onion")
+	resp := alice.RPC.SendOnion(sendReq)
+	require.True(ht, resp.Success, "expected successful onion send")
+	require.Empty(ht, resp.ErrorMessage, "unexpected failure to send onion")
+
+	// Finally, check that the Alice's payment is correctly marked
+	// succeeded.
+	//
+	// NOTE(calvin): We are not able to lookup the payment using normal
+	// means currently. I think this is because we deliver the onion
+	// directly to the switch without persisting any record via Control
+	// Tower as is done by the ChannelRouter for other payments!
+	// ht.AssertPaymentStatus(alice, preimage, lnrpc.Payment_SUCCEEDED)
+	// ht.AssertAmountPaid()
+
+	// Query for the result of the payment via onion!
+	//
+	// NOTE(calvin): This currently blocks until payment success/failure.
+	trackReq := &routerrpc.TrackOnionRequest{
+		AttemptId:   0,
+		PaymentHash: paymentHash,
+		SessionKey:  onionResp.SessionKey,
+		HopPubkeys:  onionResp.HopPubkeys,
+		// SharedSecrets: [][]byte,
```

Comment:
IIUC SharedSecrets is not covered by any itest (this test and test `testTrackOnion`).

I propose to add a test scenario testing this field.